support proxy db tunnel mfa access

GavinFrazar opened the pull request:
Closes #12538

This PR adds MFA support for `tsh proxy db --tunnel` - the basic problem with MFA is that the certs are short-lived and expire after 1 minute, which means the local proxy only worked for 1 minute before this PR.

I added middleware to `alpnproxy.LocalProxy` which is invoked when the proxy starts and then on each new connection to check/renew database certs.

The database certs are kept in memory only, in preparation for approval of RFD 90 #16739.
The main thing I've left out of this PR is adjusting the TTL for database mfa certs, as that will required RFD approval before implementing.
edit: For now, the UX improves from a broken local proxy every minute, to an MFA check every minute. The RFD will significantly improve the UX further, I'm making this PR sooner than RFD approval since the middleware is needed to unblock Teleport Connect team.

@ravicious

edit:
## UX Change
### Before
local proxy:
```
[I] [14:33:33] gavin@mac ~/work/teleport  
$ tsh proxy db --port 8080 --tunnel teleport-mysql                                                  (base) 
Tap any security key
Started authenticated tunnel for the MySQL database "teleport-mysql" in cluster "gavin-dev.cloud.gravitational.io" on 127.0.0.1:8080.

Use the following command to connect to the database:
  $ mysql --port 8080 --host localhost --protocol TCP
```

mysql:
```
[I] [14:32:36] gavin@mac ~/work/teleport  
$ echo "select 1;" | mysql -h localhost -P 8080 --protocol TCP -u teleport50
1
1
[I] [14:34:36] gavin@mac ~/work/teleport  
$ echo "select 1;" | mysql -h localhost -P 8080 --protocol TCP -u teleport50
ERROR 2013 (HY000): Lost connection to MySQL server at 'reading initial communication packet', system error: 57
```

### After:
local proxy:
```
[I] [14:31:26] gavin@mac ~/work/teleport  
$ tsh proxy db --port 8080 --tunnel teleport-mysql
Started authenticated tunnel for the MySQL database "teleport-mysql" in cluster "gavin-dev.cloud.gravitational.io" on 127.0.0.1:8080.

Use the following command to connect to the database:
  $ mysql --port 8080 --host localhost --protocol TCP
MFA is required to access database "teleport-mysql"
Tap any security key
Database certificate renewed: valid until 2022-10-05T21:32:30Z [valid for 1m0s]
MFA is required to access database "teleport-mysql"
Tap any security key
Database certificate renewed: valid until 2022-10-05T21:33:32Z [valid for 1m0s]
```

mysql:
```
[I] [14:31:41] gavin@mac ~/work/teleport  
$ echo "select 1;" | mysql -h localhost -P 8080 --protocol TCP -u teleport50
1
1
[I] [14:32:30] gavin@mac ~/work/teleport  
$ echo "select 1;" | mysql -h localhost -P 8080 --protocol TCP -u teleport50
1
1
```

And local proxy without MFA required (doesn't print a prompt context message before "Tap any security key"):
```
[I] [14:35:38] gavin@mac ~/work/teleport  
$ tsh proxy db --port 8080 --tunnel teleport-mysql
Started authenticated tunnel for the MySQL database "teleport-mysql" in cluster "gavin-dev.cloud.gravitational.io" on 127.0.0.1:8080.

Use the following command to connect to the database:
  $ mysql --port 8080 --host localhost --protocol TCP
Database certificate renewed: valid until 2022-10-06T07:27:21Z [valid for 9h52m0s]
```

OTP local proxy works too, but you need to pass an mfa auth preference (this is true for `tsh db login` as well)
```
$ tsh proxy db --port 8080 --tunnel teleport-mysql --auth=local --mfa-mode=otp
Started authenticated tunnel for the MySQL database "teleport-mysql" in cluster "gavin-dev.cloud.gravitational.io" on 127.0.0.1:8080.

Use the following command to connect to the database:
  $ mysql --port 8080 --host localhost --protocol TCP
MFA is required to access database "teleport-mysql"
Enter an OTP code from a device:
Database certificate renewed: valid until 2022-10-06T00:16:11Z [valid for 1m0s]
MFA is required to access database "teleport-mysql"
Enter an OTP code from a device:
Database certificate renewed: valid until 2022-10-06T00:17:18Z [valid for 1m0s]
```

GavinFrazar committed:
Add local proxy middleware for db cert checking

GavinFrazar committed:
Use tls conversion util instead of inline

GavinFrazar committed:
Add middleware to local proxy config

GavinFrazar committed:
Add middleware configuration in tsh

GavinFrazar committed:
Use route to database check and set defaults func

GavinFrazar committed:
Dont trigger normal db login flow if using local proxy tunnel

GavinFrazar committed:
Split out adding client creds into helper func for testing

GavinFrazar committed:
Add integration test for local proxy tunnel db cert middleware

GavinFrazar committed:
Add unit test for local proxy middleware

GavinFrazar committed:
Update comment

greedy52 commented on lib/srv/alpnproxy/middleware.go:
```diff
@@ -0,0 +1,145 @@
+	OnStart(ctx context.Context, lp *LocalProxy) error
+}
+
+// DBCertChecker is a middleware that ensures that the local proxy has valid TLS database certs.
+type DBCertChecker struct {
```
Personally, I feel the `DBCertChecker` is client specific. It will be great if we can keep this in `tsh` or `lib/client` to avoid `alpnproxy` depending on `lib/client`. What do you guys think?

smallinsky replied:
+1

GavinFrazar replied:
moved to lib/client

greedy52 commented on integration/proxy/proxy_test.go:
```diff
@@ -754,6 +755,100 @@ func TestALPNSNIProxyDatabaseAccess(t *testing.T) {
+
+		// configure local proxy without certs but with cert checking/reissuing middleware
+		// local proxy middleware should fetch a DB cert when the local proxy starts
+		lp1 := mustStartALPNLocalProxyWithConfig(t, alpnproxy.LocalProxyConfig{
```
This new test might be flaky. Since our CI can be slow, if it takes more than session time before hitting this client call here, the test can be stuck with the prompt (simulated by adding a 6 seconds delay here). Maybe we can look back at the fake clock, and/or maybe use an interface/mock for some part of the middleware?

GavinFrazar replied:
Well, I had hoped to make the integration test actually show cert renewal working instead of mocking middleware logic. I can inject the fake clock into the middleware itself, making it think that cert renewal is required and triggering the cert renewal, but what I didn't like about this is that the test would not fail with bad certs even if middleware did nothing at all - so to make it possible for such a test to fail I would need to examine the local proxy certs before/after each connection and verify that they have been changed.

I'll look at this again

greedy52 replied:
Cool. I think I am more concerned about the prompt, as long as it doesn't block when the test fails.

GavinFrazar replied:
Changed the test to just compare certs before/after mocked expiration time instead of using a short TTL cert with real waiting

greedy52 commented on lib/srv/alpnproxy/local_proxy.go:
```diff
@@ -37,6 +38,7 @@ import (
 // LocalProxy allows upgrading incoming connection to TLS where custom TLS values are set SNI ALPN and
 // updated connection is forwarded to remote ALPN SNI teleport proxy service.
 type LocalProxy struct {
+	*sync.Mutex
```
Is this intentional? Feels dangerous to expose `Lock/Unlock` to the outside world.

greedy52 commented on lib/srv/alpnproxy/local_proxy.go:
```diff
@@ -129,7 +140,14 @@ func (l *LocalProxy) Start(ctx context.Context) error {
 		}
+
 		go func() {
+			if l.cfg.Middleware != nil {
+				if err := l.cfg.Middleware.OnNewConnection(ctx, l, conn); err != nil {
+					log.WithError(err).Errorf("Failed to handle connection.")
+					return
+				}
+			}
```
I recall one discussion we had is on whether should the callback blocks or not. Some database clients may create multiple connections at the same time. I don't think our middleware is thread-safe (see #5577).

@ravicious to confirm whether Connect prefers the callback to block or not.

Alternatively, if we want both sync/async, maybe we can have both in the interface.

GavinFrazar replied:
I preferred the async method before because I wasn't just checking if the certs expired - I was checking if MFA was required for each connection and treating each connection as a new session, so it made more sense back then. I just didn't change it after we decided to only check cert expiration, and I think the UX is better with a blocking call, so I will change this to block before spawning the goroutine.

Regarding thread-safety: I don't think the async version has any undefined behavior since certs are guarded by a mutex and copied into the ALPN dialer. Suppose you have goroutines `A` and `B`: `A` checks certs and decides they don't need to be renewed, so it proceeds to handle the downstream connection. Before `A` configures its dialer, `B` checks certs and decides they need to be renewed, and then renews the certs. Now `A` and `B` will use the latest certs for their dialers.

ravicious replied:
Sorry, I didn't have time to do an actual review of the PR so I'm just responding to the mention.

> @ravicious to confirm whether Connect prefers the callback to block or not.

I'm still new to the Go world, but I think the callback can block and then the caller can decide to create a goroutine within the callback if they want the callback to become essentially nonblocking?

For a notification about an expired cert when the user attempts to connect to the db, tsh daemon is going to make a gRPC call to the server running in the Electron app. This can be completely async.

For per-session MFA in db tunnels, I haven't looked into that closely. But I imagine at some point tshd will wait for Connect to solve the MFA challenge. I'm not sure yet if this will block and the connection won't be created until the challenge is solved. But I imagine this is how it works in tsh right now with per-session MFA for stuff such as `tsh ssh`.

I'm probably of not much help here until I dive into the rest of the PR.

GavinFrazar replied:
Changed this to a blocking check before the goroutine, but I'll leave this convo unresolved if we want to discuss moving the code in the goroutine into the interface.

greedy52 commented on lib/srv/alpnproxy/middleware.go:
```diff
@@ -0,0 +1,145 @@
+		return true, nil
+	}
+	err := utils.VerifyTLSCertificateExpiry(certs[0], c.clock)
+	if err != nil && errors.As(err, &x509.CertificateInvalidError{}) {
+		log.WithError(err).Debug("need cert renewal")
+		return true, nil
```
nit. just curious if there are other reasons the certificate is invalid that may cause false positive here.

GavinFrazar replied:
I don't think any of them would be false positive, but there might be some path in `x509.ParseCertificate` that returns this error. I think I should actually trigger cert renewal for any error though, thanks for spotting this

greedy52 commented on lib/utils/certs.go:
```diff
@@ -285,4 +286,21 @@ func NewCertPoolFromPath(path string) (*x509.CertPool, error) {
 	return pool, nil
 }
 
+func TLSCertToX509(cert tls.Certificate) (*x509.Certificate, error) {
```
nit: go doc

greedy52 commented on tool/tsh/db.go:
```diff
@@ -813,6 +807,13 @@ func getDatabase(cf *CLIConf, tc *client.TeleportClient, dbName string) (types.D
 }
 
 func needDatabaseRelogin(cf *CLIConf, tc *client.TeleportClient, database *tlsca.RouteToDatabase, profile *client.ProfileStatus) (bool, error) {
+	if cf.LocalProxyTunnel {
+		// Don't login to database here if local proxy tunnel is enabled.
+		// When local proxy tunnel is enabled, the local proxy will check if DB login is needed when
+		// it starts and on each new connection.
+		return false, nil
+	}
+
```
Does the local proxy OnStart check if re-login is needed because of different `--db-user` or `--db-name`?

GavinFrazar replied:
I don't think different --db-user or --db-name will be possible, because the local proxy tunnel will have certs in memory each time. I suppose the `--cert-file` and `--key-file` could be passed though, so in that case I can add cert subject checking

greedy52 replied:
ah, ok. so the local proxy is configured with zero certs and `OnStart` will always fetch a new one? So there is some performance penalty that we are refetching certs even though it might not be necessary (e.g. non MFA flows). I don't think it's a big deal, but if it's something easy we can do to optimize the performance that would be great.

GavinFrazar replied:
Prior behavior would attempt to load certs from disk if they were already there - I don't see anything wrong with adding that back. You're right that it's an extra cert fetch in a non-MFA flow currently. But I will keep this change for `needDatabaseRelogin` so proxy tunnels don't trigger a "normal" login, because it's simpler logic to keep track of.

greedy52 commented on tool/tsh/db.go:
```diff
@@ -640,12 +632,13 @@ type localProxyConfig struct {
 func prepareLocalProxyOptions(arg *localProxyConfig) (localProxyOpts, error) {
-	// If user requested no client auth, open an authenticated tunnel using
-	// client cert/key of the database.
 	certFile := arg.cliConf.LocalProxyCertFile
 	keyFile := arg.cliConf.LocalProxyKeyFile
```
not sure what's the history on these two flags (`--cert-file`, `--key-file`). do we still need them?

GavinFrazar replied:
I don't know what the history of them is. If they provide cert and key files, the local proxy tunnel will use them until they expire, at which point local proxy middleware will try to renew certs in memory

GavinFrazar commented on lib/srv/alpnproxy/middleware.go:
```diff
@@ -0,0 +1,145 @@
+		accessRequests = profile.ActiveRequests.AccessRequests
+	}
+	var key *client.Key
+	if err := client.RetryWithRelogin(ctx, c.tc, func() error {
+		newKey, err := c.tc.IssueUserCertsWithMFA(ctx, client.ReissueParams{
```
Good suggestion! I will add context messages, as you are right the user does not have good context, With a command like `tsh db connect`, it's obvious why they are being prompted, but not when the prompt can happen after some idle time with ` tsh proxy db --tunnel`

GavinFrazar replied:
How about just:
```
Local proxy tunnel requires credentials to access database X
Tap any security key:
Local proxy tunnel credentials successfully renewed
...
Local proxy tunnel credentials for database X have expired, attempting to renew
Tap any security key:
Local proxy tunnel credentials successfully renewed
...
Local proxy tunnel credentials for database X have expired, attempting to renew
Enter password for Teleport user gavin:
Tap any security key:
Tap any security key:
Local proxy tunnel credentials successfully renewed
```

This way I don't have to make the extra API call to auth server to see if MFA is required - I just check cert expiration.

smallinsky replied:
Looks fine to me but to ensure the `Local proxy tunnel requires credentials to access database ` will be printed only for MFA flow ?

GavinFrazar commented on lib/srv/alpnproxy/local_proxy.go:
```diff
@@ -129,7 +140,14 @@ func (l *LocalProxy) Start(ctx context.Context) error {
+
 		go func() {
+			if l.cfg.Middleware != nil {
+				if err := l.cfg.Middleware.OnNewConnection(ctx, l, conn); err != nil {
```
There is no data race because the local proxy certs protected by a mutex, but I think the UX would be better if we moved this to the main loop.

GavinFrazar committed:
Make middleware on new conn block

GavinFrazar committed:
godoc

GavinFrazar committed:
Make any cert check error trigger cert renewal in local proxy middleware

GavinFrazar committed:
Move dbcertchecker into lib/client

GavinFrazar committed:
Remove unneeded mutex in local proxy and unused func in lib/utils

GavinFrazar committed:
Make local proxy middleware integration test more robust

GavinFrazar committed:
Print message before mfa prompt in proxy tunnel

GavinFrazar committed:
Add before prompt option to test

GavinFrazar commented:
@greedy52 @smallinsky this is ready I think.

I've addressed your feedback regarding adding some context before prompting the user for MFA and I've made the integration test more robust using a fake clock to mock cert expiration.

GavinFrazar committed:
Remove unneeded comment

smallinsky commented:
@GavinFrazar 
In case of normal flow without MFA can we skip renewal of database certs if certs was already generated and are valid ? 
For instance: 
The certs generated by `tsh db connect --db-user=marek --db-name=postgres postgres01` 
```
➜  tsh git:(gavinfrazar/proxy_db_tunnel_mfa_access) ✗ ./tsh db connect --db-user=marek --db-name=postgres postgres01
psql (14.4, server 13.4)
SSL connection (protocol: TLSv1.3, cipher: TLS_AES_128_GCM_SHA256, bits: 128, compression: off)
Type "help" for help.
```
can be reused for tsh proxy db though in current flow always renews the certs: 
```
postgres=> ^D\q
➜  tsh git:(gavinfrazar/proxy_db_tunnel_mfa_access) ✗ ./tsh proxy db postgres01 --tunnel  -p 1111
Started authenticated tunnel for the PostgreSQL database "postgres01" in cluster "root" on 127.0.0.1:1111.

Use the following command to connect to the database:
  $ psql postgres://marek@localhost:1111/postgres

Local proxy tunnel requires credentials to access database "postgres01"
Local proxy tunnel credentials successfully renewed
```

Also the: 

``Local proxy tunnel requires credentials to access database "postgres01"` log entry  might be a bit confusing for a user.  I would keep this more like a debug log because no user action is needed in this case. 

In case of MFA flow right now following message is printed: 
```
local proxy tunnel requires credentials to access database "teleport-mysql"
Enter an OTP code from a device:
```
Could we emphasise in the log entry that local proxy that this is MFA flow ? 

@xinding33 Could you take a look at UX changes ?

GavinFrazar committed:
Change local proxy messages to be more clear

GavinFrazar committed:
Pass local proxy opts by reference

GavinFrazar committed:
Pass certs in opts instead of cert/key file path

This is so we can check if the error is recoverable while preparing
local proxy options. A tunneled local proxy can ignore the error because
it does not rely on cert files - it can just renew its certs if
necessary.

GavinFrazar commented:
@smallinsky

> In case of normal flow without MFA can we skip renewal of database certs if certs was already generated and are valid ?

Done.

> Could we emphasize in the log entry that local proxy that this is MFA flow ?

Done. See:
```
MFA is required to access database "teleport-mysql"
Tap any security key
Database certificate renewed: valid until 2022-10-06T18:23:53Z [valid for 1m0s]
```

GavinFrazar commented on lib/tlsca/ca.go:
```diff
@@ -218,9 +219,25 @@ type RouteToDatabase struct {
+	// be encoded into it. This is required to be able to tell which database
+	// user to authenticate the connection as.
+	if r.Protocol == defaults.ProtocolMongoDB && r.Username == "" {
+		return trace.BadParameter("please provide the database user name using --db-user flag")
```
Yeah I wasn't very satisfied with putting this here either. I think I'll just make that a function in tsh and check the db route before creating the middleware

GavinFrazar committed:
Move db route checking back to tsh

smallinsky reviewed (approved):
Linter seems to fail on `a = a + 1` rule but otherwise LGTM.

GavinFrazar commented on lib/client/mfa.go:
```diff
@@ -50,6 +50,10 @@ func (p *mfaPrompt) PromptPIN() (string, error) {
 
 // PromptMFAChallengeOpts groups optional settings for PromptMFAChallenge.
 type PromptMFAChallengeOpts struct {
+	// BeforePrompt is an optional method to print before the prompt.
```
Typo, it was supposed to be "an optional message". I can change it to a callback since that's more flexible though. Maybe call it "AdviceBefore" like emacs lisp terminology lol

GavinFrazar committed:
Fix lint err

GavinFrazar committed:
Merge branch 'master' into gavinfrazar/proxy_db_tunnel_mfa_access

GavinFrazar committed:
Fix typo and print the hint to same writer as the mfa prompt

GavinFrazar committed:
Merge branch 'master' into gavinfrazar/proxy_db_tunnel_mfa_access

GavinFrazar committed:
Merge branch 'master' into gavinfrazar/proxy_db_tunnel_mfa_access

ravicious commented on lib/srv/alpnproxy/local_proxy.go:
```diff
@@ -129,6 +146,14 @@ func (l *LocalProxy) Start(ctx context.Context) error {
 		}
+
+		if l.cfg.Middleware != nil {
+			if err := l.cfg.Middleware.OnNewConnection(ctx, l, conn); err != nil {
+				log.WithError(err).Errorf("Middleware failed to handle new connection.")
+				continue
+			}
```
I'm currently implementing a mini version of what you have in `local_proxy_middleware.go`, a middleware that specifically handles expired certs only (no MFA support yet) and reaches out to Connect to trigger the login modal.

If `l.cfg.Middleware.OnNewConnection` returns an error, from the client's perspective the connection just hangs indefinitely. Should we close the connection on error instead?

GavinFrazar replied:
The reason I did continue was because the connection is passed into the middleware to decide what to do. Although I did forget to close the connection on error in the middleware...

GavinFrazar replied:
So I agree we should update this, either to close connection here or to handle the error by closing the connection in the middleware

ravicious replied:
I think we should just close it here unless we find a reason to not close the connection if the middleware encounters an error.

---

BTW, for now the middleware I'm adding will be super simple as it'll only check for the expired cert and nothing more. In the future though, once we get around to finally adding per-session MFA support to Connect, I plan to refactor the existing middleware a bit so that it can be shared between Connect and tsh.

From what I gathered, the logic will be mostly the same and the two will only differ (as usual) in how they handle UI interactions (CLI vs GUI).
